rs/nervous_system/candid_utils/src/validation/tests.rs

```diff
@@ -0,0 +1,138 @@
+                    f : (byte, int, nat, int8) -> (List);
+                    g : (List) -> (int) query;
+                }
+            "#,
+            "(record {}, (11 : nat32))",
+            Ok(()),
+        ),
+        (
+            "Complex service with two arguments (missing 1st arg)",
+            r#"
+                type List = opt record { head: int; tail: List };
+                type byte = nat8;
+                service : (x : record { foo : opt record {} }, y : nat32) -> {
+                    f : (byte, int, nat, int8) -> (List);
+                    g : (List) -> (int) query;
+                }
+            "#,
+            "((11 : nat32))",
+            Err(()),
+        ),
+        (
+            "Complex service with two arguments (missing 2nd arg)",
+            r#"
+                type List = opt record { head: int; tail: List };
+                type byte = nat8;
+                service : (x : record { foo : opt record {} }, y : nat32) -> {
+                    f : (byte, int, nat, int8) -> (List);
+                    g : (List) -> (int) query;
+                }
+            "#,
+            "(record {})",
+            Err(()),
+        ),
+        (
+            "Complex service with two arguments (missing both args)",
+            r#"
+                type List = opt record { head: int; tail: List };
+                type byte = nat8;
+                service : (x : record { foo : opt record {} }, y : nat32) -> {
+                    f : (byte, int, nat, int8) -> (List);
+                    g : (List) -> (int) query;
+                }
+            "#,
+            "()",
+            Err(()),
+        ),
+        (
+            "Trivial service with two arguments (wrong arg order)",
+            r#"
+                service : (x : record { foo : opt record {} }, y : nat32) -> {
+                    g : () -> (int) query;
+                }
+            "#,
+            "((11 : nat32), record {})",
+            Err(()),
+        ),
+        (
+            "Trivial service with one record argument (subtyping holds)",
+            r#"
+                service : (x : record { foo : opt nat; bar : opt nat }) -> {
+                    g : () -> (int) query;
+                }
+            "#,
+            "(record { foobar = opt (1984 : nat); foo = opt (42 : nat) })",
+            Ok(()),
+        ),
+        (
+            "Trivial service with one record argument (extra field breaks subtyping)",
+            r#"
+                service : (x : record { foo : opt nat; bar : nat }) -> {
+                    g : () -> (int) query;
+                }
+            "#,
+            "(record { foobar = (1984 : nat); foo = opt (42 : nat) })",
+            Err(()),
+        ),
+    ] {
+        let observed_result =
+            validate_upgrade_args(candid_service.to_string(), upgrade_arg.to_string());
+
+        match (observed_result, expected_result) {
```

Comment:
No change required.

```rust
assert_eq!(observed_result.is_ok(), expected_result.is_ok(), "{:?}", observed_result);
```

Reply:
Tried this, but it doesn't print errors as nicely as I want.